src/analysis/vector/geometry_checker/qgssinglegeometrycheck.h

```diff
@@ -0,0 +1,143 @@
+/***************************************************************************
+                      qgssinglegeometrycheck.h
+                     --------------------------------------
+Date                 : 6.9.2018
+Copyright            : (C) 2018 by Matthias Kuhn
+email                : matthias@opengis.ch
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#ifndef QGSSINGLEGEOMETRYCHECK_H
+#define QGSSINGLEGEOMETRYCHECK_H
+
+#define SIP_NO_FILE
+
+#include <QList>
+#include <QCoreApplication>
+
+#include "qgsgeometry.h"
+#include "qgsgeometrycheck.h"
+
+#include "qgis_analysis.h"
+
+class QgsFeature;
+class QgsSingleGeometryCheck;
+
+/**
+ * An error from a QgsSingleGeometryCheck.
+ *
+ */
+class ANALYSIS_EXPORT QgsSingleGeometryCheckError
+{
+  public:
+    QgsSingleGeometryCheckError( const QgsSingleGeometryCheck *check, const QgsGeometry &geometry, const QgsPoint &errorLocation, const QgsVertexId &vertexId = QgsVertexId() )
+      : mCheck( check )
+      , mGeometry( geometry )
+      , mErrorLocation( errorLocation )
+      , mVertexId( vertexId )
+    {}
+
+    virtual ~QgsSingleGeometryCheckError() = default;
+
+    virtual void update( const QgsSingleGeometryCheckError *other );
+    virtual bool isEqual( const QgsSingleGeometryCheckError *other ) const;
+    virtual bool handleChanges( const QList<QgsGeometryCheck::Change> &changes );
+    virtual QString description() const;
+
+    /**
+     * The check that created this error.
+     *
+     * \since QGIS 3.4
+     */
+    const QgsSingleGeometryCheck *check() const;
+
+    /**
+     * The exact location of the error.
+     *
+     * \since QGIS 3.4
+     */
+    QgsPoint errorLocation() const;
+
+    /**
+     * The vertex id of the error. May be invalid depending on the check.
+     *
+     * \since QGIS 3.4
+     */
+    QgsVertexId vertexId() const;
+
+  protected:
+    const QgsSingleGeometryCheck *mCheck;
+    QgsGeometry mGeometry;
+    QgsPoint mErrorLocation;
```

Comment:
Should this support multiple locations? Eg self intersection multiple times

Reply:
I was more wondering if it should support polygons (overlap etc) it's a pure port of what's existing

Reply:
Yeah - that's a very good point. Maybe this should just be a QgsGeometry for error location, so multipoint, lines, polygons, etc are supported